lxd/patches.go

```diff
@@ -245,58 +247,150 @@ func patchStorageApi(name string, d *Daemon) error {
 	// is supposed to be created on.
-	defaultID, defaultProfile, err := dbProfileGet(d.db, "default")
+	profiles, err := dbProfiles(d.db)
 	if err == nil {
-		foundRoot := false
-		for k, v := range defaultProfile.Devices {
-			if v["type"] == "disk" && v["path"] == "/" && v["source"] == "" {
-				// Add the default storage pool.
-				defaultProfile.Devices[k]["pool"] = poolName
-				foundRoot = true
+		for _, pName := range profiles {
+			pID, p, err := dbProfileGet(d.db, pName)
+			if err != nil {
+				// Do not fail.
+				continue
 			}
-		}
 
-		if !foundRoot {
-			rootDev := map[string]string{}
-			rootDev["type"] = "disk"
-			rootDev["path"] = "/"
-			rootDev["pool"] = poolName
-			if defaultProfile.Devices == nil {
-				defaultProfile.Devices = map[string]map[string]string{}
+			// Check for a root disk device entry
+			k, _ := containerGetRootDiskDevice(p.Devices)
+			if k != "" {
+				// If found, set the pool property
+				p.Devices[k]["pool"] = poolName
+			} else if k == "" && pName == "default" {
+				// The default profile should have a valid root
+				// disk device entry.
+				rootDev := map[string]string{}
+				rootDev["type"] = "disk"
+				rootDev["path"] = "/"
+				rootDev["pool"] = poolName
+				if p.Devices == nil {
+					p.Devices = map[string]map[string]string{}
+				}
+
+				// Make sure that we do not overwrite a device the user
+				// is currently using under the name "root".
+				rootDevName := "root"
+				for i := 0; i < 100; i++ {
+					if p.Devices[rootDevName] == nil {
+						break
+					}
+					rootDevName = fmt.Sprintf("root%d", i)
+					continue
+				}
+				p.Devices["root"] = rootDev
+			}
+
+			// This is nasty, but we need to clear the profiles config and
+			// devices in order to add the new root device including the
+			// newly added storage pool.
+			tx, err := dbBegin(d.db)
+			if err != nil {
+				return err
+			}
+
+			err = dbProfileConfigClear(tx, pID)
+			if err != nil {
+				shared.LogWarnf("Could not clear old profile configuration.")
+				tx.Rollback()
+				continue
+			}
+
+			err = dbProfileConfigAdd(tx, pID, p.Config)
+			if err != nil {
+				shared.LogWarnf("Could not add new profile configuration.")
+				tx.Rollback()
+				continue
+			}
+
+			err = dbDevicesAdd(tx, "profile", pID, p.Devices)
+			if err != nil {
+				shared.LogWarnf("Could not add new profile devices.")
+				tx.Rollback()
+				continue
+			}
+
+			err = tx.Commit()
+			if err != nil {
+				shared.LogWarnf("Could not commit database transaction.")
+				tx.Rollback()
+				continue
 			}
-			defaultProfile.Devices["root"] = rootDev
 		}
+	}
 
-		// This is nasty, but we need to clear the profiles config and
-		// devices in order to add the new root device including the
-		// newly added storage pool.
-		tx, err := dbBegin(d.db)
+	// When the either no default profile is not detected or some containers
+	// do not use the default profile, give these containers a valid local
+	// root disk device with a pool property set.
```

Comment:
Sounds good. The comment currently in the file isn't valid English, has double negative, ... :)